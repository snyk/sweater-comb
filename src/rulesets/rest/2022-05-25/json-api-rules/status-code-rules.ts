import { Ruleset, RuleError, ResponseRule } from "@useoptic/rulesets-base";
import { links } from "../../../../docs";
import { isOpenApiPath, isBatchPostOperation } from "../utils";

const valid4xxCodes = new ResponseRule({
  name: "valid 4xx status codes",
  matches: (response) => response.statusCode.startsWith("4"),
  rule: (responseAssertions) => {
    const allowed4xxStatusCodes = ["400", "401", "403", "404", "409", "429"];
    responseAssertions.added(
      "support the correct 4xx status codes",
      (response) => {
        if (!allowed4xxStatusCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected response to not support status code ${response.statusCode}`,
          });
        }
      },
    );

    responseAssertions.changed(
      "support the correct 4xx status codes",
      (beforeResponse, response) => {
        if (!allowed4xxStatusCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected response to not support status code ${response.statusCode}`,
          });
        }
      },
    );
  },
});

const delete2xxCodes = new ResponseRule({
  name: "valid 2xx status codes for delete",
  matches: (response, rulesContext) =>
    response.statusCode.startsWith("2") &&
    rulesContext.operation.method === "delete",
  rule: (responseAssertions) => {
    const validDelete2xxCodes = ["200", "204"];
    responseAssertions.added(
      "support the correct 2xx status codes",
      (response) => {
        if (!validDelete2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected response to not support status code ${response.statusCode}`,
          });
        }
      },
    );

    responseAssertions.changed(
      "support the correct 2xx status codes",
      (beforeResponse, response) => {
        if (!validDelete2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected response to not support status code ${response.statusCode}`,
          });
        }
      },
    );
  },
});

const post2xxCodes = new ResponseRule({
  name: "valid 2xx status codes for post",
  matches: (response, rulesContext) =>
    !isBatchPostOperation(rulesContext.operation.requests) &&
    response.statusCode.startsWith("2") &&
    rulesContext.operation.method === "post",
  rule: (responseAssertions) => {
    const validPost2xxCodes = ["201"];
    responseAssertions.added(
      "support the correct 2xx status codes",
      (response) => {
        if (!validPost2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected POST response to only support 201, not ${response.statusCode}`,
          });
        }
      },
    );

    responseAssertions.changed(
      "support the correct 2xx status codes",
      (beforeResponse, response) => {
        if (!validPost2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected POST response to only support 201, not ${response.statusCode}`,
          });
        }
      },
    );
  },
});

const get2xxCodes = new ResponseRule({
  name: "valid 2xx status codes for get",
  matches: (response, rulesContext) =>
    response.statusCode.startsWith("2") &&
    rulesContext.operation.method === "get",
  rule: (responseAssertions) => {
    responseAssertions.added(
      "support the correct 2xx status codes",
      (response) => {
        if (response.statusCode !== "200") {
          throw new RuleError({
            message: `expected GET response to only support 200, not ${response.statusCode}`,
          });
        }
      },
    );

    responseAssertions.changed(
      "support the correct 2xx status codes",
      (beforeResponse, response) => {
        if (response.statusCode !== "200") {
          throw new RuleError({
            message: `expected GET response to only support 200, not ${response.statusCode}`,
          });
        }
      },
    );
  },
});

const batchPost2xxCodes = new ResponseRule({
  name: "valid 2xx status codes for post",
  matches: (response, rulesContext) =>
    isBatchPostOperation(rulesContext.operation.requests) &&
    response.statusCode.startsWith("2") &&
    rulesContext.operation.method === "post",
  rule: (responseAssertions) => {
    const validPost2xxCodes = ["204"];
    responseAssertions.added(
      "support the correct 2xx status codes",
      (response) => {
        if (!validPost2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected POST response for batches to only support 204, not ${response.statusCode}`,
          });
        }
      },
    );

    responseAssertions.changed(
      "support the correct 2xx status codes",
      (beforeResponse, response) => {
        if (!validPost2xxCodes.includes(response.statusCode)) {
          throw new RuleError({
            message: `expected POST response for batches to only support 204, not ${response.statusCode}`,
          });
        }
      },
    );
  },
});

export const statusCodesRules = new Ruleset({
  name: "JSON:API status codes",
  docsLink: links.standards.statusCodes,
  matches: (rulesContext) => !isOpenApiPath(rulesContext.operation.path),
  rules: [
    valid4xxCodes,
    delete2xxCodes,
    post2xxCodes,
    get2xxCodes,
    batchPost2xxCodes,
  ],
});
