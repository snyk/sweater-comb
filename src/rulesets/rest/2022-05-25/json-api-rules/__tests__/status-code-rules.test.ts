import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "../../__tests__/fixtures";
import { statusCodesRules } from "../status-code-rules";

const baseJson = TestHelpers.createEmptySpec();

describe("status code rules", () => {
  test("fails when an invalid 4xx status code is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users/{user_id}": {
          get: {
            responses: {
              "405": {
                description: "i am not valid",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test("fails when an invalid delete 2xx code is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users/{user_id}": {
          delete: {
            responses: {
              "201": {
                description: "i am not valid",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test("fails when an invalid post 2xx code is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users/{user_id}": {
          post: {
            responses: {
              "200": {
                description: "i am not valid",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test("fails when an invalid get 2xx code is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users/{user_id}": {
          get: {
            responses: {
              "203": {
                description: "i am not valid",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          name: "valid 2xx status codes for get",
          error: "expected GET response to only support 200, not 203",
        }),
      ]),
    );
  });

  test("fails when an invalid batch post 2xx code is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users": {
          post: {
            requestBody: {
              content: {
                "application/vnd.api+json": {
                  schema: {
                    type: "object",
                    properties: {
                      data: {
                        type: "array",
                        items: {},
                      },
                    },
                  },
                },
              },
            },
            responses: {
              "200": {
                description: "i am not valid",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results.filter((result) => !result.passed)).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          error:
            "expected POST response for batches to only support 204, not 200",
        }),
      ]),
    );
    expect(results).toMatchSnapshot();
  });

  test("fails when an invalid batch post request is specified", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users/{user_id}": {
          post: {
            requestBody: {
              content: {
                "application/vnd.api+json": {
                  schema: {
                    type: "array",
                    items: {},
                  },
                },
              },
            },
            responses: {
              "204": {
                description: "i am not valid",
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results.filter((result) => !result.passed)).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          error: "expected POST response to only support 201, not 204",
        }),
      ]),
    );
    expect(results).toMatchSnapshot();
  });

  test("passes for a valid batch post 204 code", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users": {
          post: {
            requestBody: {
              content: {
                "application/vnd.api+json": {
                  schema: {
                    type: "object",
                    properties: {
                      data: {
                        type: "array",
                        items: {
                          type: "object",
                          properties: {
                            id: "some-id",
                            type: "some-type",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
            responses: {
              "204": {
                description: "got it",
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([statusCodesRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(true);
    expect(results).toMatchSnapshot();
  });
});
