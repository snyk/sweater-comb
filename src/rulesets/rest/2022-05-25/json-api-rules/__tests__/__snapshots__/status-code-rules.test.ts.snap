// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`status code rules fails when an invalid 4xx status code is specified 1`] = `
[
  {
    "change": {
      "added": {
        "description": "i am not valid",
        "statusCode": "405",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "405",
          },
          "method": "get",
          "path": "/api/users/{user_id}",
        },
        "conceptualPath": [
          "operations",
          "/api/users/{}",
          "get",
          "responses",
          "405",
        ],
        "jsonPath": "/paths/~1api~1users~1{user_id}/get/responses/405",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": "expected response to not support status code 405",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 4xx status codes",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users/{user_id} response 405",
  },
]
`;

exports[`status code rules fails when an invalid batch post 2xx code is specified 1`] = `
[
  {
    "change": {
      "added": {
        "description": "i am not valid",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "200",
          },
          "method": "post",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "post",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1users/post/responses/200",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": "expected POST response for batches to only support 204, not 200",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 2xx status codes for post",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "POST /api/users response 200",
  },
]
`;

exports[`status code rules fails when an invalid batch post request is specified 1`] = `
[
  {
    "change": {
      "added": {
        "description": "i am not valid",
        "statusCode": "204",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "204",
          },
          "method": "post",
          "path": "/api/users/{user_id}",
        },
        "conceptualPath": [
          "operations",
          "/api/users/{}",
          "post",
          "responses",
          "204",
        ],
        "jsonPath": "/paths/~1api~1users~1{user_id}/post/responses/204",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": "expected POST response to only support 201, not 204",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 2xx status codes for post",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "POST /api/users/{user_id} response 204",
  },
]
`;

exports[`status code rules fails when an invalid delete 2xx code is specified 1`] = `
[
  {
    "change": {
      "added": {
        "description": "i am not valid",
        "statusCode": "201",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "201",
          },
          "method": "delete",
          "path": "/api/users/{user_id}",
        },
        "conceptualPath": [
          "operations",
          "/api/users/{}",
          "delete",
          "responses",
          "201",
        ],
        "jsonPath": "/paths/~1api~1users~1{user_id}/delete/responses/201",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": "expected response to not support status code 201",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 2xx status codes for delete",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "DELETE /api/users/{user_id} response 201",
  },
]
`;

exports[`status code rules fails when an invalid post 2xx code is specified 1`] = `
[
  {
    "change": {
      "added": {
        "description": "i am not valid",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "200",
          },
          "method": "post",
          "path": "/api/users/{user_id}",
        },
        "conceptualPath": [
          "operations",
          "/api/users/{}",
          "post",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1users~1{user_id}/post/responses/200",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": "expected POST response to only support 201, not 200",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 2xx status codes for post",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "POST /api/users/{user_id} response 200",
  },
]
`;

exports[`status code rules passes for a valid batch post 204 code 1`] = `
[
  {
    "change": {
      "added": {
        "description": "got it",
        "statusCode": "204",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "204",
          },
          "method": "post",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "post",
          "responses",
          "204",
        ],
        "jsonPath": "/paths/~1api~1users/post/responses/204",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#status-codes",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid 2xx status codes for post",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "POST /api/users response 204",
  },
]
`;
